### PR TITLE
Introduce os_fd_t and windows port of os_sys_calls_impl

### PR DESCRIPTION
**Pull request by wrowe**
Description:
- Add os_fd_t as the new data type for os_sys_calls socket methods
  necessary to port to windows, where SOCKET is 2x width of an int
- Add SysCallSocketResult of that data type
- Normalize and "const-ify" all usage of iovec and msghdr types
- Add additional os_sys_calls methods all required for porting
  the various test/ sources to Windows in subsequent PRs
- Introduce the win32 port of os_sys_calls_impl

This patch defers the actual work of employing the os_sys_calls API
where needed for Windows, which is a much more involved PR review.

Risk Level: Moderate
Testing: local on linux gcc, windows msvc
Docs Changes: n/a
Release Notes: n/a

### CODE REVIEW

File: source/common/api/win32/os_sys_calls_impl.cc [hunk 1]
```diff
@@ -0,0 +1,334 @@
...
+  const int rc = ::ioctlsocket(sockfd, request, static_cast<u_long*>(argp));
+  return {rc, rc != -1 ? 0 : ::WSAGetLastError()};
+}
+
+SysCallIntResult OsSysCallsImpl::close(os_fd_t fd) {
+  const int rc = ::closesocket(fd);
+  return {rc, rc != -1 ? 0 : ::WSAGetLastError()};
+}
+
+using WSABUFPtr = std::unique_ptr<WSABUF[]>;
```
[the remark this answers is not in the document]
Comment (sunjayBhatia) [reply source/common/api/win32/os_sys_calls_impl.cc]: sounds good, will move

**Comment by wrowe** [comment]
> One question I have is whether our existing tests are sufficient to exercise all the glue code in win32/os_sys_calls_impl.cc

It wasn't, nor was the existing posix implementation. That patch is coming in the next PR.

File: source/common/network/address_impl.cc [hunk 2]
```diff
@@ -166,8 +166,8 @@ IoHandlePtr InstanceBase::socketFromSocketType(SocketType socket_type) const {
     domain = AF_UNIX;
   }
 
-  const Api::SysCallIntResult result = Api::OsSysCallsSingleton::get().socket(domain, flags, 0);
-  RELEASE_ASSERT(result.rc_ != -1,
+  const Api::SysCallSocketResult result = os_sys_calls_.socket(domain, flags, 0);
```
[the remark this answers is not in the document]
Comment (wrowe) [reply source/common/network/address_impl.cc]: For classes which are broad consumers, this style is preferable IMO. If it's a one-off, the singleton lookup approach is less complex and preferable, but we are open to either approach.

**Comment by sunjayBhatia** [comment]
> /retest

test passes locally but failed in CI, just retesting to make sure

File: source/common/api/win32/os_sys_calls_impl.cc [hunk 3]
```diff
@@ -0,0 +1,334 @@
...
+
+  return std::move(wsa_msg);
+}
+
+SysCallSizeResult OsSysCallsImpl::recvmsg(os_fd_t sockfd, msghdr* msg, int flags) {
+  LPFN_WSARECVMSG WSARecvMsgFnPtr = NULL;
+  DWORD bytes_received;
+  if (NULL == (WSARecvMsgFnPtr = GetWSARecvMsgFnPtr())) {
+    PANIC("WSARecvMsg has not been implemented by this socket provider");
+  }
```
[the remark this answers is not in the document]
Comment (wrowe) [reply source/common/api/win32/os_sys_calls_impl.cc]: Resolved in the next commit

File: source/common/api/win32/os_sys_calls_impl.cc [hunk 4]
```diff
@@ -0,0 +1,334 @@
...
+  return {bytes_received, 0};
+}
+
+SysCallSizeResult OsSysCallsImpl::recv(os_fd_t socket, void* buffer, size_t length, int flags) {
+  const ssize_t rc = ::recv(socket, static_cast<char*>(buffer), length, flags);
+  return {rc, rc != -1 ? 0 : ::WSAGetLastError()};
+}
+
+static LPFN_WSARECVMSG GetWSARecvMsgFnPtr() {
```
[the remark this answers is not in the document]
Comment (wrowe) [reply source/common/api/win32/os_sys_calls_impl.cc]: Moving discussion to #envoy-windows-dev and reporting back...

[the remark this answers is not in the document]
Comment (wrowe) [reply source/common/api/win32/os_sys_calls_impl.cc]: Confirmed this latest push addresses the concern.

File: source/common/network/address_impl.cc (hunk 2 above)
[the remark this answers is not in the document]
Comment (sunjayBhatia) [reply source/common/network/address_impl.cc]: Resolved in https://github.com/envoyproxy/envoy/pull/10036/commits/5f78ad3ae1d1dd8734df078891a08d36a169155f

**Comment by sunjayBhatia** [comment]
In our efforts to try to resolve the circle CI failure, we noticed we could reduce the relevant test file and did so in: https://github.com/envoyproxy/envoy/pull/10036/commits/6b5d982b611c852850bc334263a1bca55a32db8d

File: source/common/api/win32/os_sys_calls_impl.cc [hunk 5]
```diff
@@ -0,0 +1,333 @@
...
+namespace {
+
+using WSABUFPtr = std::unique_ptr<WSABUF[]>;
+
+struct wsabufResult {
+  DWORD num_vec_;
+  WSABUFPtr wsabuf_;
+};
+
+static wsabufResult iovecToWSABUF(const iovec* vec, int in_vec) {
```
Comment (lizan) [inline source/common/api/win32/os_sys_calls_impl.cc]: style nit: no need of `static` inside anonymous namespace (ditto below)

File: source/common/api/win32/os_sys_calls_impl.cc [hunk 6]
```diff
@@ -0,0 +1,333 @@
...
+  wsabufResult wsabuf = iovecToWSABUF(msg->msg_iov, msg->msg_iovlen);
+  wsa_msg->lpBuffers = wsabuf.wsabuf_.get();
+  wsa_msg->dwBufferCount = wsabuf.num_vec_;
+  WSABUF control;
+  control.buf = reinterpret_cast<CHAR*>(msg->msg_control);
+  control.len = msg->msg_controllen;
+  wsa_msg->Control = control;
+  wsa_msg->dwFlags = msg->msg_flags;
+
+  return std::move(wsa_msg);
```
Comment (lizan) [inline source/common/api/win32/os_sys_calls_impl.cc]: nit: redundant std::move.

File: source/common/api/win32/os_sys_calls_impl.cc [hunk 7]
```diff
@@ -0,0 +1,333 @@
...
+}
+
+SysCallSizeResult OsSysCallsImpl::recv(os_fd_t socket, void* buffer, size_t length, int flags) {
+  const ssize_t rc = ::recv(socket, static_cast<char*>(buffer), length, flags);
+  return {rc, rc != -1 ? 0 : ::WSAGetLastError()};
+}
+
+SysCallSizeResult OsSysCallsImpl::recvmsg(os_fd_t sockfd, msghdr* msg, int flags) {
+  DWORD bytes_received;
+  LPFN_WSARECVMSG WSARecvMsgFnPtr = getFnPtrWSARecvMsg();
```
Comment (lizan) [inline source/common/api/win32/os_sys_calls_impl.cc]: style: variable should be snake_case.

Comment (wrowe) [reply source/common/api/win32/os_sys_calls_impl.cc]: Patch incoming, thank you for the review

**Comment by sunjayBhatia** [comment]
retesting due to coverage tests segfault (same as https://github.com/envoyproxy/envoy/issues/10012)

/retest

**Comment by sunjayBhatia** [comment]
retesting (again) due to coverage tests segfault (same as https://github.com/envoyproxy/envoy/issues/10012)

/retest

**Comment by sunjayBhatia** [comment]
@zuercher looks like we just keep failing on https://github.com/envoyproxy/envoy/issues/10012, should we continue to retest until it passes? a previous commit before some windows-only changes had this pass (see: https://github.com/envoyproxy/envoy/pull/10036/commits, specifically commit https://github.com/envoyproxy/envoy/pull/10036/commits/5f78ad3ae1d1dd8734df078891a08d36a169155f)

**Comment by mattklein123** [comment]
We can merge w/o coverage passing until we fix it. cc @lizan

**Pull request merged**